{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from aerobot.io import load_training_data, load_validation_data\n",
    "from aerobot.utls import process_data\n",
    "from aerobot.models import LogisticClassifier, GeneralClassifier\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A model based only on oxidation state features\n",
    "\n",
    "The oxidation state of DSDNA is an analytic function of its GC content. For proteins and RNA, we actually need to calculate these values, so we load a reference table. \n",
    "\n",
    "The code below loads amino acid and DNA sequence content for each genome, calculates the nominal oxidation states of those, and then uses them as features in a 3 class predictor of oxygen utilization. \n",
    "\n",
    "TODO: need to get access to coding sequences to calculate ZC of RNA coding seqs. Josh is on this. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "CANONICAL_NTS = ['A', 'C', 'G', 'T']\n",
    "GC_NTS = ['G', 'C']\n",
    "\n",
    "def calc_gc(nt1_features_df):\n",
    "    \"\"\"Calculate the GC content of the canonical nucleotides.\n",
    "\n",
    "    Args:\n",
    "        nt1_features_df (pd.DataFrame): A dataframe containing the counts of\n",
    "            single nucleotides.\n",
    "\n",
    "    Returns:\n",
    "        pd.Series: A series containing the GC content of each genome.\n",
    "    \"\"\"\n",
    "    canonical_data = nt1_features_df[CANONICAL_NTS]\n",
    "    gc = canonical_data[GC_NTS].sum(axis=1) / canonical_data.sum(axis=1)\n",
    "    gc.name = 'gc_content'\n",
    "    return gc\n",
    "\n",
    "\n",
    "AA_NOSC_DF = pd.read_csv('aa_nosc.csv', index_col=0)\n",
    "AA_NC = AA_NOSC_DF.NC\n",
    "AA_ZC = AA_NOSC_DF.NOSC\n",
    "CANONICAL_AAS = AA_NOSC_DF.index.tolist()\n",
    "\n",
    "def calc_aa_zc(aa1_features_df):\n",
    "    \"\"\"Calculate the formal C oxidation state.\n",
    "\n",
    "    Args:\n",
    "        aa1_features_df (pd.DataFrame): A dataframe containing the counts of\n",
    "            single amino acids.\n",
    "\n",
    "    Returns:\n",
    "        pd.Series: The formal C oxidation state.\n",
    "    \"\"\"\n",
    "    aas = sorted(set(aa1_features_df.columns).intersection(CANONICAL_AAS))\n",
    "    canonical_data = aa1_features_df[aas]\n",
    "    NC_total = (canonical_data @ AA_NC[aas])\n",
    "    Ne_total = (canonical_data @ (AA_ZC[aas] * AA_NC[aas]))\n",
    "    mean_cds_zc = Ne_total / NC_total\n",
    "    mean_cds_zc.name = 'cds_aa_zc'\n",
    "\n",
    "    cols = [mean_cds_zc]\n",
    "    for elt in 'CNOS':\n",
    "        col = 'N{0}'.format(elt)\n",
    "        ns = AA_NOSC_DF[col]\n",
    "        totals = (canonical_data @ ns[aas])\n",
    "        means = totals / canonical_data.sum(axis=1)\n",
    "        means.name = 'cds_aa_{0}'.format(col)\n",
    "        cols.append(means)\n",
    "    return pd.concat(cols, axis=1)\n",
    "\n",
    "\n",
    "NT_NOSC_DF = pd.read_csv('nt_nosc.csv')\n",
    "RNA_NOSC_DF = NT_NOSC_DF[NT_NOSC_DF.type == 'RNA']\n",
    "RNA_NOSC_DF = RNA_NOSC_DF.set_index('letter_code')\n",
    "RNA_NC = RNA_NOSC_DF.NC\n",
    "RNA_ZC = RNA_NOSC_DF.NOSC\n",
    "# Includes DNA and RNA names\n",
    "CANONICAL_NTS_ALL = RNA_NOSC_DF.index.unique().tolist()\n",
    "\n",
    "def calc_rna_zc(nt1_features_df):\n",
    "    \"\"\"Calculate the formal C oxidation state of an RNA.\n",
    "\n",
    "    Args:\n",
    "        nt1_features_df (pd.DataFrame): A dataframe containing the counts of\n",
    "            single nucleotides, either RNA or DNA. Assumed to single stranded.\n",
    "\n",
    "    Returns:\n",
    "        pd.Series: The formal C oxidation state of the RNA coding sequences.\n",
    "    \"\"\"\n",
    "    my_nts = sorted(set(nt1_features_df.columns).intersection(CANONICAL_NTS_ALL))\n",
    "    canonical_data = nt1_features_df[my_nts]\n",
    "    NC_total = (canonical_data @ RNA_NC[my_nts])\n",
    "    Ne_total = (canonical_data @ (RNA_ZC[my_nts] * RNA_NC[my_nts]))\n",
    "    mean_cds_zc = Ne_total / NC_total\n",
    "    mean_cds_zc.name = 'cds_nt_zc'\n",
    "\n",
    "    cols = [mean_cds_zc]\n",
    "    for elt in 'CNOS':\n",
    "        col = 'N{0}'.format(elt)\n",
    "        ns = RNA_NOSC_DF[col]\n",
    "        totals = (canonical_data @ ns[my_nts])\n",
    "        means = totals / canonical_data.sum(axis=1)\n",
    "        means.name = 'cds_nt_{0}'.format(col)\n",
    "        cols.append(means)\n",
    "    return pd.concat(cols, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading metadata...\n",
      "Loading nt data...\n",
      "Calculating GC content...\n",
      "Loading cds data...\n",
      "Loading aa data...\n",
      "Calculating protein ZC...\n",
      "Merging features...\n"
     ]
    }
   ],
   "source": [
    "print(\"Loading metadata...\")\n",
    "\n",
    "feature_type = \"metadata\"\n",
    "training_data_metadata = load_training_data(feature_type=feature_type)\n",
    "validation_data_metadata = load_validation_data(feature_type=feature_type)\n",
    "\n",
    "train_ngenes = training_data_metadata[\"features\"][\"number_of_genes\"]\n",
    "test_ngenes = validation_data_metadata[\"features\"][\"number_of_genes\"]\n",
    "\n",
    "print(\"Loading nt data...\")\n",
    "\n",
    "feature_type = \"nt_1mer\"\n",
    "training_data_nt = load_training_data(feature_type=feature_type)\n",
    "validation_data_nt = load_validation_data(feature_type=feature_type)\n",
    "\n",
    "print(\"Calculating GC content...\")\n",
    "train_gc = calc_gc(training_data_nt[\"features\"])\n",
    "test_gc = calc_gc(validation_data_nt[\"features\"])\n",
    "\n",
    "print(\"Loading cds data...\")\n",
    "feature_type = \"cds_1mer\"\n",
    "training_data_cds_nt = load_training_data(feature_type=feature_type)\n",
    "validation_data_cds_nt = load_validation_data(feature_type=feature_type)\n",
    "\n",
    "train_rna_zc = calc_rna_zc(training_data_cds_nt[\"features\"])\n",
    "test_rna_zc = calc_rna_zc(validation_data_cds_nt[\"features\"])\n",
    "\n",
    "# Strip the .x suffixes from the test index\n",
    "test_rna_zc.index = [a for a,b in test_rna_zc.index.str.split('.')]\n",
    "\n",
    "print(\"Loading aa data...\")\n",
    "feature_type = 'aa_1mer'\n",
    "training_data_aa = load_training_data(feature_type=feature_type)\n",
    "validation_data_aa = load_validation_data(feature_type=feature_type)\n",
    "\n",
    "print(\"Calculating protein ZC...\")\n",
    "train_aa_zc = calc_aa_zc(training_data_aa[\"features\"])\n",
    "test_aa_zc = calc_aa_zc(validation_data_aa[\"features\"])\n",
    "\n",
    "print(\"Merging features...\")\n",
    "training_features = pd.merge(train_gc, train_aa_zc, left_index=True, right_index=True)\n",
    "validation_features = pd.merge(test_gc, test_aa_zc, left_index=True, right_index=True)\n",
    "training_features = pd.merge(training_features, train_rna_zc, left_index=True, right_index=True)\n",
    "validation_features = pd.merge(validation_features, test_rna_zc, left_index=True, right_index=True)\n",
    "training_features = pd.merge(training_features, train_ngenes, left_index=True, right_index=True)\n",
    "validation_features = pd.merge(validation_features, test_ngenes, left_index=True, right_index=True)\n",
    "\n",
    "if training_features['gc_content'].size == 0:\n",
    "    print(\"No training features found!\")\n",
    "if validation_features['gc_content'].size == 0:\n",
    "    print(\"No validation features found!\")\n",
    "\n",
    "training_data = training_data_aa\n",
    "validation_data = validation_data_aa\n",
    "training_data[\"features\"] = training_features\n",
    "validation_data[\"features\"] = validation_features\n",
    "\n",
    "cleaned_data = process_data(\n",
    "    training_data[\"features\"], training_data[\"labels\"][\"physiology\"],\n",
    "    validation_data[\"features\"], validation_data[\"labels\"][\"physiology\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gc_content</th>\n",
       "      <th>cds_aa_zc</th>\n",
       "      <th>cds_aa_NC</th>\n",
       "      <th>cds_aa_NN</th>\n",
       "      <th>cds_aa_NO</th>\n",
       "      <th>cds_aa_NS</th>\n",
       "      <th>cds_nt_zc</th>\n",
       "      <th>cds_nt_NC</th>\n",
       "      <th>cds_nt_NN</th>\n",
       "      <th>cds_nt_NO</th>\n",
       "      <th>cds_nt_NS</th>\n",
       "      <th>number_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000007325.1</th>\n",
       "      <td>0.271516</td>\n",
       "      <td>-0.223904</td>\n",
       "      <td>5.143151</td>\n",
       "      <td>1.312394</td>\n",
       "      <td>2.503998</td>\n",
       "      <td>0.031280</td>\n",
       "      <td>0.906262</td>\n",
       "      <td>9.580959</td>\n",
       "      <td>3.843651</td>\n",
       "      <td>8.011308</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000008085.1</th>\n",
       "      <td>0.315595</td>\n",
       "      <td>-0.261506</td>\n",
       "      <td>5.313145</td>\n",
       "      <td>1.337274</td>\n",
       "      <td>2.473235</td>\n",
       "      <td>0.025246</td>\n",
       "      <td>0.908078</td>\n",
       "      <td>9.579899</td>\n",
       "      <td>3.869975</td>\n",
       "      <td>8.023750</td>\n",
       "      <td>0.0</td>\n",
       "      <td>583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000009845.1</th>\n",
       "      <td>0.277599</td>\n",
       "      <td>-0.225050</td>\n",
       "      <td>5.268872</td>\n",
       "      <td>1.365388</td>\n",
       "      <td>2.493303</td>\n",
       "      <td>0.028975</td>\n",
       "      <td>0.881118</td>\n",
       "      <td>9.527548</td>\n",
       "      <td>3.735673</td>\n",
       "      <td>8.080170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000010085.1</th>\n",
       "      <td>0.545049</td>\n",
       "      <td>-0.147696</td>\n",
       "      <td>4.929883</td>\n",
       "      <td>1.381895</td>\n",
       "      <td>2.430102</td>\n",
       "      <td>0.040638</td>\n",
       "      <td>0.908349</td>\n",
       "      <td>9.511738</td>\n",
       "      <td>3.802405</td>\n",
       "      <td>8.265712</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000010565.1</th>\n",
       "      <td>0.529611</td>\n",
       "      <td>-0.179736</td>\n",
       "      <td>4.949605</td>\n",
       "      <td>1.357630</td>\n",
       "      <td>2.428669</td>\n",
       "      <td>0.036650</td>\n",
       "      <td>0.917596</td>\n",
       "      <td>9.536937</td>\n",
       "      <td>3.859736</td>\n",
       "      <td>8.220832</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_904846235.1</th>\n",
       "      <td>0.427827</td>\n",
       "      <td>-0.150589</td>\n",
       "      <td>4.925863</td>\n",
       "      <td>1.330677</td>\n",
       "      <td>2.485576</td>\n",
       "      <td>0.036234</td>\n",
       "      <td>0.899661</td>\n",
       "      <td>9.519435</td>\n",
       "      <td>3.768315</td>\n",
       "      <td>8.197186</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_904846285.1</th>\n",
       "      <td>0.424173</td>\n",
       "      <td>-0.152190</td>\n",
       "      <td>4.936632</td>\n",
       "      <td>1.334622</td>\n",
       "      <td>2.483212</td>\n",
       "      <td>0.035871</td>\n",
       "      <td>0.899430</td>\n",
       "      <td>9.518972</td>\n",
       "      <td>3.765110</td>\n",
       "      <td>8.197720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_904848585.1</th>\n",
       "      <td>0.619125</td>\n",
       "      <td>-0.136393</td>\n",
       "      <td>4.828909</td>\n",
       "      <td>1.365579</td>\n",
       "      <td>2.412853</td>\n",
       "      <td>0.033587</td>\n",
       "      <td>0.912379</td>\n",
       "      <td>9.505566</td>\n",
       "      <td>3.824078</td>\n",
       "      <td>8.306658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_905187535.1</th>\n",
       "      <td>0.397108</td>\n",
       "      <td>-0.184541</td>\n",
       "      <td>5.055416</td>\n",
       "      <td>1.320646</td>\n",
       "      <td>2.502971</td>\n",
       "      <td>0.028774</td>\n",
       "      <td>0.898044</td>\n",
       "      <td>9.526067</td>\n",
       "      <td>3.764989</td>\n",
       "      <td>8.166035</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_907163115.1</th>\n",
       "      <td>0.637897</td>\n",
       "      <td>-0.142745</td>\n",
       "      <td>4.892210</td>\n",
       "      <td>1.378553</td>\n",
       "      <td>2.429864</td>\n",
       "      <td>0.032745</td>\n",
       "      <td>0.911327</td>\n",
       "      <td>9.501178</td>\n",
       "      <td>3.827814</td>\n",
       "      <td>8.317148</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4276</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3342 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gc_content  cds_aa_zc  cds_aa_NC  cds_aa_NN  cds_aa_NO  \\\n",
       "GCA_000007325.1    0.271516  -0.223904   5.143151   1.312394   2.503998   \n",
       "GCA_000008085.1    0.315595  -0.261506   5.313145   1.337274   2.473235   \n",
       "GCA_000009845.1    0.277599  -0.225050   5.268872   1.365388   2.493303   \n",
       "GCA_000010085.1    0.545049  -0.147696   4.929883   1.381895   2.430102   \n",
       "GCA_000010565.1    0.529611  -0.179736   4.949605   1.357630   2.428669   \n",
       "...                     ...        ...        ...        ...        ...   \n",
       "GCF_904846235.1    0.427827  -0.150589   4.925863   1.330677   2.485576   \n",
       "GCF_904846285.1    0.424173  -0.152190   4.936632   1.334622   2.483212   \n",
       "GCF_904848585.1    0.619125  -0.136393   4.828909   1.365579   2.412853   \n",
       "GCF_905187535.1    0.397108  -0.184541   5.055416   1.320646   2.502971   \n",
       "GCF_907163115.1    0.637897  -0.142745   4.892210   1.378553   2.429864   \n",
       "\n",
       "                 cds_aa_NS  cds_nt_zc  cds_nt_NC  cds_nt_NN  cds_nt_NO  \\\n",
       "GCA_000007325.1   0.031280   0.906262   9.580959   3.843651   8.011308   \n",
       "GCA_000008085.1   0.025246   0.908078   9.579899   3.869975   8.023750   \n",
       "GCA_000009845.1   0.028975   0.881118   9.527548   3.735673   8.080170   \n",
       "GCA_000010085.1   0.040638   0.908349   9.511738   3.802405   8.265712   \n",
       "GCA_000010565.1   0.036650   0.917596   9.536937   3.859736   8.220832   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "GCF_904846235.1   0.036234   0.899661   9.519435   3.768315   8.197186   \n",
       "GCF_904846285.1   0.035871   0.899430   9.518972   3.765110   8.197720   \n",
       "GCF_904848585.1   0.033587   0.912379   9.505566   3.824078   8.306658   \n",
       "GCF_905187535.1   0.028774   0.898044   9.526067   3.764989   8.166035   \n",
       "GCF_907163115.1   0.032745   0.911327   9.501178   3.827814   8.317148   \n",
       "\n",
       "                 cds_nt_NS  number_of_genes  \n",
       "GCA_000007325.1        0.0             2022  \n",
       "GCA_000008085.1        0.0              583  \n",
       "GCA_000009845.1        0.0              956  \n",
       "GCA_000010085.1        0.0             5742  \n",
       "GCA_000010565.1        0.0             3001  \n",
       "...                    ...              ...  \n",
       "GCF_904846235.1        0.0             2708  \n",
       "GCF_904846285.1        0.0             2748  \n",
       "GCF_904848585.1        0.0             7761  \n",
       "GCF_905187535.1        0.0             1941  \n",
       "GCF_907163115.1        0.0             4276  \n",
       "\n",
       "[3342 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_data[\"features\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting model...\n",
      "Accuracy: 0.7570505287896592\n",
      "Balanced Accuracy: 0.580434883626373\n",
      "Test Accuracy: 0.6106870229007634\n",
      "Test Balanced Accuracy: 0.5331186258722491\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Aerobe</th>\n",
       "      <th>Anaerobe</th>\n",
       "      <th>Facultative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Aerobe</th>\n",
       "      <td>0.887292</td>\n",
       "      <td>0.108683</td>\n",
       "      <td>0.004025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anaerobe</th>\n",
       "      <td>0.153972</td>\n",
       "      <td>0.842752</td>\n",
       "      <td>0.003276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Facultative</th>\n",
       "      <td>0.720721</td>\n",
       "      <td>0.268018</td>\n",
       "      <td>0.011261</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Aerobe  Anaerobe  Facultative\n",
       "Aerobe       0.887292  0.108683     0.004025\n",
       "Anaerobe     0.153972  0.842752     0.003276\n",
       "Facultative  0.720721  0.268018     0.011261"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Fitting model...\")\n",
    "model = LogisticClassifier(max_iter=100000, normalize=True, random_state=None)\n",
    "model.fit(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "accuracy = model.score(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "balanced_accuracy = model.balanced_accuracy(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "\n",
    "print(\"Accuracy: \" + str(accuracy))\n",
    "print(\"Balanced Accuracy: \" + str(balanced_accuracy))\n",
    "\n",
    "# compute accuracy and balanced accuracy on test set\n",
    "test_accuracy = model.score(cleaned_data[\"X_test\"], cleaned_data[\"y_test\"])\n",
    "test_balanced_accuracy = model.balanced_accuracy(cleaned_data[\"X_test\"], cleaned_data[\"y_test\"])\n",
    "print(\"Test Accuracy: \" + str(test_accuracy))\n",
    "print(\"Test Balanced Accuracy: \" + str(test_balanced_accuracy))\n",
    "\n",
    "C = model.confusion_matrix(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "C = pd.DataFrame(C,index=model.classifier.classes_, columns=model.classifier.classes_)\n",
    "C.apply(lambda x: x/x.sum(), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting two-class model...\n",
      "Accuracy: 0.8572267920094007\n",
      "Balanced Accuracy: 0.8390593983814323\n",
      "Test Accuracy: 0.7633587786259542\n",
      "Test Balanced Accuracy: 0.768293404318249\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Aerobe</th>\n",
       "      <th>Anaerobe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Aerobe</th>\n",
       "      <td>0.903344</td>\n",
       "      <td>0.096656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anaerobe</th>\n",
       "      <td>0.225225</td>\n",
       "      <td>0.774775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Aerobe  Anaerobe\n",
       "Aerobe    0.903344  0.096656\n",
       "Anaerobe  0.225225  0.774775"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make it a two-class thing and run again\n",
    "training_data[\"labels\"][\"physiology\"] = training_data[\"labels\"][\"physiology\"].replace(\"Facultative\", \"Aerobe\")\n",
    "validation_data[\"labels\"][\"physiology\"] = validation_data[\"labels\"][\"physiology\"].replace(\"Facultative\", \"Aerobe\")\n",
    "\n",
    "cleaned_data = process_data(\n",
    "    training_data[\"features\"], training_data[\"labels\"][\"physiology\"],\n",
    "    validation_data[\"features\"], validation_data[\"labels\"][\"physiology\"])\n",
    "\n",
    "\n",
    "print(\"Fitting two-class model...\")\n",
    "model = LogisticClassifier(max_iter=100000, normalize=True, random_state=None)\n",
    "model.fit(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "accuracy = model.score(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "balanced_accuracy = model.balanced_accuracy(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "\n",
    "print(\"Accuracy: \" + str(accuracy))\n",
    "print(\"Balanced Accuracy: \" + str(balanced_accuracy))\n",
    "\n",
    "# compute accuracy and balanced accuracy on test set\n",
    "test_accuracy = model.score(cleaned_data[\"X_test\"], cleaned_data[\"y_test\"])\n",
    "test_balanced_accuracy = model.balanced_accuracy(cleaned_data[\"X_test\"], cleaned_data[\"y_test\"])\n",
    "print(\"Test Accuracy: \" + str(test_accuracy))\n",
    "print(\"Test Balanced Accuracy: \" + str(test_balanced_accuracy))\n",
    "\n",
    "C = model.confusion_matrix(cleaned_data[\"X_train\"], cleaned_data[\"y_train\"])\n",
    "C = pd.DataFrame(C,index=model.classifier.classes_, columns=model.classifier.classes_)\n",
    "C.apply(lambda x: x/x.sum(), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aerobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
